### DISCUSSION TITLE
How to use `protoc-gen-validate` (PGV) with `connect-go` ?

### DISCUSSION THREAD

**Discussion by s-takehana** [connectrpc/connect-go#426] [Q&A] [answered]
How to use [`protoc-gen-validate` (PGV)](https://github.com/bufbuild/protoc-gen-validate) with `connect-go` ?

I want to call `Validate` / `ValidateAll` in [interceptor](https://connect.build/docs/go/interceptors) func.

**Comment by s-takehana** [comment, marked as answer]
```go
type validator interface {
	ValidateAll() error
}

func NewValidationInterceptor() connect.UnaryInterceptorFunc {
	interceptor := func(next connect.UnaryFunc) connect.UnaryFunc {
		return connect.UnaryFunc(func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
			if v, ok := req.Any().(validator); ok {
				if err := v.ValidateAll(); err != nil {
					return nil, connect.NewError(connect.CodeInvalidArgument, err)
				}
			}
			return next(ctx, req)
		})
	}
	return connect.UnaryInterceptorFunc(interceptor)
}
```

**Reply by akshayjshah** [reply]
For unary RPCs, this looks correct to me! To support streaming RPCs, you'd also need to write wrappers for `connect.StreamingClientConn` and `connect.StreamingHandlerConn`.

**Reply by s-takehana** [reply]
Thanks!

**Reply by ymmt2005** [reply]
@akshayjshah 
May I ask a question?
It seems that the streaming interceptor can't handle incoming messages,
so we can't implement an interceptor that validates all incoming messages.
Am I correct?

**Reply by akshayjshah** [reply]
Nope, you definitely can validate messages - the interceptor would wrap `StreamingHandlerConn` or `StreamingClientConn`, and validate each time a message is sent or received.

**Reply by ymmt2005** [reply]
Understood. Thank you!

**Reply by ymmt2005** [reply]
I ended up with something like
```go
type interceptor struct{}

func (i *interceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
	return func(ctx context.Context, conn connect.StreamingHandlerConn) error {
		return next(ctx, validatingHandlerConn{conn})
	}
}

type validatingHandlerConn struct {
	connect.StreamingHandlerConn
}

func (conn validatingHandlerConn) Receive(a any) error {
	if err := conn.StreamingHandlerConn.Receive(a); err != nil {
		return err
	}
	// call Validate or ValidateAll ...
	return nil
}
```

**Reply by akshayjshah** [reply]
That looks about right!

**Reply by s-takehana** [reply]
[validate-go](https://github.com/connectrpc/validate-go) is useful for `protovalidate` users.
